recipes/zlib-bytesread-to-byteswritten/src/workflow.ts

```diff
@@ -0,0 +1,134 @@
+];
+
+const FUNC_KINDS = [
+	'function_declaration',
+	'function_expression',
+	'arrow_function',
+] as const;
+
+// Helper to find all identifier nodes within a given node
+function findIdentifiers(node: ReturnType<SgRoot<Js>['root']>): string[] {
+	const identifiers: string[] = [];
+	const stack = [node];
+
+	while (stack.length) {
+		const current = stack.pop()!;
+		if (current.kind() === 'identifier') {
+			identifiers.push(current.text());
+		}
+		stack.push(...current.children());
+	}
+
+	return identifiers;
+}
+
+export default function transform(root: SgRoot<Js>): string | null {
+	const rootNode = root.root();
+	const edits: Edit[] = [];
+	const linesToRemove: Range[] = [];
+
+	// 1 Find all static and dynamic zlib imports/requires
+	const importNodes = [
+		...getNodeRequireCalls(root, 'node:zlib'),
+		...getNodeImportStatements(root, 'node:zlib'),
+		...getNodeImportCalls(root, 'node:zlib'),
+	];
+
+	const factoryBindings = new Set<string>();
+	const streamVariables: string[] = [];
+
+	// 1.a Resolve all local bindings from "node:zlib"
+	for (const node of importNodes) {
+		for (const factory of ZLIB_FACTORIES) {
+			const binding = resolveBindingPath(node, `$.${factory}`);
+			if (binding) factoryBindings.add(binding);
+		}
+	}
+
+	// If no import is found that means we can skip transformation on this file
+	if (!importNodes.length) return null;
+
+	// 2 Track variables assigned from factories (const, let, var)
+	for (const binding of factoryBindings) {
+		const matches = rootNode.findAll({
+			rule: {
+				kind: 'variable_declarator',
+				has: {
+					field: 'value',
+					kind: 'call_expression',
+					pattern: `${binding}($$$ARGS)`,
+				},
+			},
+		});
+
+		for (const match of matches) {
+			const varMatch = match.field('name');
+
+			if (varMatch) {
+				const varName = varMatch.text();
+				if (!streamVariables.includes(varName)) streamVariables.push(varName);
+			}
+		}
+	}
+
+	// 3 Replace .bytesRead → .bytesWritten for tracked variables
+	for (const variable of streamVariables) {
+		const matches = rootNode.findAll({
+			rule: { pattern: `${variable}.bytesRead` },
+		});
+
+		for (const match of matches) {
+			edits.push(
+				match.replace(match.text().replace('.bytesRead', '.bytesWritten')),
+			);
+		}
+	}
+
+	// Step 4: Replace .bytesRead → .bytesWritten for function parameters
+	for (const kind of FUNC_KINDS) {
+		const funcs = rootNode.findAll({ rule: { kind } });
+
+		for (const func of funcs) {
+			const formalParamsNode = func
+				.children()
+				.find((child) => child.kind() === 'formal_parameters');
+
+			if (!formalParamsNode) continue;
+
+			const paramNames = findIdentifiers(formalParamsNode);
+
+			for (const paramName of paramNames) {
+				const matches = rootNode.findAll({
+					rule: { pattern: `${paramName}.bytesRead` },
+				});
+				for (const match of matches) {
+					edits.push(
+						match.replace(match.text().replace('.bytesRead', '.bytesWritten')),
+					);
+				}
+			}
+		}
+	}
```

Comment:
I think the find Identifier util can re removed if we simplify the way the function parameters are fetched.

I created a patch for this one as well (and create a new testing scenario with a function that receive more than one parameter).

https://gist.githubusercontent.com/brunocroh/71cdfb6cc9ee4cfcd8e690ab9f888b3f/raw/25bb2e3b71e57f136ef7f81afda058d87986c4e2/gistfile0.txt

```suggestion
	for (const kind of FUNC_KINDS) {
		const funcs = rootNode.findAll({ rule: { kind } });

		for (const func of funcs) {
			const paramNames = func
				.field('parameters')
				?.findAll({ rule: { kind: 'identifier' } });

			for (const paramName of paramNames) {
				// replace member_expressions that use ${paramName}.bytesRead inside the function context
				const matches = func.findAll({
					rule: {
						kind: 'member_expression',
						pattern: `${paramName.text()}.bytesRead`,
					},
				});
				for (const match of matches) {
					edits.push(
						match.replace(match.text().replace('.bytesRead', '.bytesWritten')),
					);
				}
			}
		}
	}
```